llvm-hs-pure/src/LLVM/AST/Metadata.hs

```diff
@@ -0,0 +1,105 @@
+module LLVM.AST.Metadata where
+
+import LLVM.Prelude
+import LLVM.AST.Name
+
+data MDNode
+  = DIExpression -- nyi
+  | DIGlobalVariableExpression -- nyi
+  | DILocation Word32 Word32 DILocalScope
+  | DIMacroNode -- nyi
+  | DINode DINode
+  | MDTuple -- nyi
+  deriving (Eq, Ord, Read, Show, Typeable, Data, Generic)
+
+data DINode
+  = DIEnumerator Word32 Name
+  | DIImportedEntity Word32 Name DIScope DINode {- ? -} Word32
+  | DIObjCProperty Word32 Word32 Name DIFile Name Name DIType
+  | DIScope DIScope
+  | DISubrange Word32 Word32
+  | DITemplateParameter DITemplateParameter
+  | DIVariable DIVariable
+  | GenericDINode -- idk yet
+  deriving (Eq, Ord, Read, Show, Typeable, Data, Generic)
+
+data DIScope
+  = DICompileUnit DIFile ShortByteString Bool ShortByteString Word32 ShortByteString Word32 MDNode MDNode MDNode MDNode MDNode Word64
```

Comment:
Please use record types for these types. Figuring out what the individual parameters are supposed to be will otherwise be really hard.

Reply:
I agree they are more understandable but what should I do about partial record selectors? Ignore that and assume consumers will build / destruct things directly? What about overlapping selectors (ie most subclasses will have a `scope` field).

Reply:
For now, I would just go with partial record selectors. While they are definitely ugly we already use them in other places (not a deliberate design decision on my part, we inherited them from `llvm-general`) and nobody has complained so far. The alternative would be to extract parameters to separate types, e.g., make a separate `DICompileUnit` type but then naming constructors becomes annoying. If someone has a good suggestion for a naming scheme here, that would also be a decent option. Maybe just drop the `DI` from the type that actually has the parameters? Either way, this should be relatively easy to change later so I wouldn’t worry too much about it.

With regardes to name collisions, we are going to have to go with prefixes. We’ve only just dropped support for `7.8` and I don’t want to drop support for `7.10` at this point so `DuplicateRecordFields` is not an option. That said, iirc GHC 8.4 is supposed to be released in February so since we follow we the 3 release policy, we could then revisit this topic.

Reply:
👍 I figured this project followed the 3-release policy. I'll probably incrementally document the constructors. It's already hard for me to keep them straight in my head.

Reply:
Depending on the usecase it could be easy to add some "classy-lens" style classes for the most common fields (ie scope, file)